### PR TITLE
Translating 04-efficiency to pytorch

### PR DESCRIPTION
**Pull request by danishpruthi**
+ translation of 04-efficiency to pytorch
+ a small fix in 04-efficiency dynet version to not update during dev.

### CODE REVIEW

**Comment by danishpruthi** [comment]
@MysteryVaibhav: please have a look once.

File: 04-efficiency-pytorch/wordemb-skip-ns.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import torch
+import torch.nn.functional as F
+
+
+class WordEmbSkip(torch.nn.Module):
+    def __init__(self, nwords, emb_size):
+        super(WordEmbSkip, self).__init__()
+
+        """ word embeddings """
+        self.word_embedding = torch.nn.Embedding(nwords, emb_size)
```
Comment (MysteryVaibhav) [inline 04-efficiency-pytorch/wordemb-skip-ns.py]: Usually while learning word embeddings, we can expect the size of matrix to be in the order of 200k - 500k. I think in those scenarios, it is useful to set the sparse computation of gradients to True. torch.nn.Embedding(nwords, emb_size, sparse=True) will make the code much faster. However, it might not be very useful here.

File: 04-efficiency-pytorch/wordemb-skip-ns.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        # initialize the weights with xavier uniform (Glorot, X. & Bengio, Y. (2010))
+        torch.nn.init.xavier_uniform_(self.word_embedding.weight)
+        """ context embeddings"""
+        self.context_embedding = torch.nn.Embedding(nwords, emb_size)
```
Comment (MysteryVaibhav) [inline 04-efficiency-pytorch/wordemb-skip-ns.py]: Can add the sparse=True flag here as well.

**Comment by MysteryVaibhav** [comment]
@danishpruthi The code for wordemb-skip-ns.py looks fine. I didn't run both the versions to compare the similarity in performance, I guess you would have done that already.
Also, did you miss converting wordemb-skip-binary.py ? Or you left it intentionally because the negative sampling one is more relevant.

**Comment by danishpruthi** [comment]
@MysteryVaibhav: Thanks for the suggestion, I made the required change. In the interest of time, I skipped `wordemb-skip-binary.py` (as the other one is more prominent). I would translate that too, and once I do it, I will add a separate PR for that. 

@neubig: please merge this.
